File: frontend/src/types/field.ts
/* ------------------------------------------------------------------ *
 * field.ts — TypeScript types mirroring the backend Pydantic models  *
 * Keep in sync with: backend/app/models/*.py                         *
 * ------------------------------------------------------------------ */

// ---------------------------------------------------------------------------
// Primitives
// ---------------------------------------------------------------------------

export type FieldType =
  | "text"
  | "tel"
  | "email"
  | "date"
  | "time"
  | "number"
  | "textarea"
  | "url"
  | "file"
  | "checkbox";

// ---------------------------------------------------------------------------
// Configuration types
// ---------------------------------------------------------------------------

export interface CustomKeywordRule {
  keyword: string;
  type: FieldType;
}

export interface FieldSchema {
  key: string;
  source_header: string;
  label: string;
  type: FieldType;
  required: boolean;
  order: number;
  column_index: number;
  placeholder?: string;
}

// ---------------------------------------------------------------------------
// API response types
// ---------------------------------------------------------------------------

/** Returned by POST /api/preview */
export interface PreviewResponse {
  spreadsheet_id: string;
  sheet_url: string;
  spreadsheet_title: string;
  worksheet_name: string;
  form_title: string;
  fields: FieldSchema[];
  custom_keywords: CustomKeywordRule[];
  warnings: string[];
}

/** Returned by POST /api/forms */
export interface CreateFormResponse {
  id: string;
  edit_token: string;
  form_url: string;
  edit_url: string;
}

/** Returned by POST /api/sheet/create */
export interface CreateSheetResponse {
  spreadsheet_id: string;
  sheet_url: string;
  worksheet_name: string;
}

export interface FormLibraryItem {
  id: string;
  form_title: string;
  sheet_url: string;
  spreadsheet_id: string;
  worksheet_name: string | null;
  field_count: number;
  submission_count: number;
  updated_at: string;
  form_url: string;
  edit_url: string;
}

export interface FormLibraryResponse {
  items: FormLibraryItem[];
}

/** Returned by GET /api/forms/:id */
export interface PublicFormResponse {
  id: string;
  form_title: string;
  description: string;
  worksheet_name: string | null;
  fields: FieldSchema[];
  autofill_columns: string[];
  ui_config?: Record<string, any> | null;
}

/** Returned by GET /api/forms/:id/edit?token=… */
export interface EditFormResponse {
  id: string;
  sheet_url: string;
  spreadsheet_id: string;
  worksheet_name: string | null;
  form_title: string;
  description: string;
  fields: FieldSchema[];
  custom_keywords: CustomKeywordRule[];
  autofill_columns: string[];
  ui_config?: Record<string, any> | null;
}

/** Returned by POST /api/forms/:id/submit */
export interface SubmitFormResponse {
  success: boolean;
  updated_range: string | null;
  timestamp: string;
}

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

/** Shape of a non-2xx JSON body from the backend */
export interface ApiError {
  detail: string;
}

File: backend/app/models/field.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, field_validator


FieldType = Literal["text", "tel", "email", "date", "time", "number", "textarea", "url", "file", "checkbox"]


class CustomKeywordRule(BaseModel):
    keyword: str = Field(..., min_length=1, max_length=80)
    type: FieldType = "text"

    @field_validator("keyword")
    @classmethod
    def clean_keyword(cls, value: str) -> str:
        cleaned = " ".join(value.strip().lower().split())
        if not cleaned:
            raise ValueError("Keyword cannot be blank")
        return cleaned


class FieldSchema(BaseModel):
    key: str
    source_header: str
    label: str
    type: FieldType = "text"
    required: bool = False
    order: int
    column_index: int
    placeholder: str | None = None

    @field_validator("key")
    @classmethod
    def clean_key(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Field key cannot be blank")
        return cleaned

    @field_validator("label")
    @classmethod
    def clean_label(cls, value: str) -> str:
        cleaned = " ".join(value.strip().split())
        if not cleaned:
            raise ValueError("Field label cannot be blank")
        return cleaned
